manticore/platforms/evm_world_state.py

```diff
@@ -0,0 +1,468 @@
+# written.
+
+
+class Storage:
+    def __init__(self, constraints: ConstraintSet, address: int):
+        """
+        :param constraints: the ConstraintSet with which this Storage object is associated
+        :param address: the address that owns this storage
+        """
+        self._data = constraints.new_array(
+            index_bits=256,
+            value_bits=256,
+            name=f"STORAGE_DATA_{address:x}",
+            avoid_collisions=True,
+            # sam.moelius: The use of default here creates unnecessary if-then-elses.  See
+            # ArrayProxy.get in expression.py.
+            # default=0,
+        )
+
+    def __getitem__(self, offset: Union[int, BitVec]) -> Union[int, BitVec]:
+        return self.get(offset, 0)
+
+    def get(self, offset: Union[int, BitVec], default: Union[int, BitVec]) -> Union[int, BitVec]:
+        if not isinstance(default, BitVec):
+            default = BitVecConstant(256, default)
+        return self._data.get(offset, default)
+
+    def set(self, offset: Union[int, BitVec], value: Union[int, BitVec]):
+        self._data[offset] = value
+
+    def get_items(self) -> List[Tuple[Union[int, BitVec], Union[int, BitVec]]]:
+        items = []
+        array = self._data.array
+        while not isinstance(array, ArrayVariable):
+            items.append((array.index, array.value))
+            array = array.array
+        return items
+
+    def dump(self, stream: TextIOBase, state: State):
+        concrete_indexes = set()
+        for sindex in self._data.written:
+            concrete_indexes.add(state.solve_one(sindex, constrain=True))
+
+        for index in concrete_indexes:
+            stream.write(
+                f"storage[{index:x}] = {state.solve_one(self._data[index], constrain=True):x}\n"
+            )
+
+
+####################################################################################################
+
+
+class WorldState:
+    def new_storage(
+        self, constraints: ConstraintSet, address: int, items: Optional[Dict[int, int]] = None
+    ) -> Storage:
+        storage = Storage(constraints, address)
+        if items is not None:
+            for key, value in items.items():
+                storage.set(key, value)
+        return storage
+
+    @abstractmethod
+    def is_remote(self) -> bool:
+        pass
+
+    @abstractmethod
+    def accounts(self) -> Set[int]:
+        pass
+
+    @abstractmethod
+    def get_nonce(self, address: int) -> Union[int, BitVec]:
+        pass
+
+    @abstractmethod
+    def get_balance(self, address: int) -> Union[int, BitVec]:
+        pass
+
+    @abstractmethod
+    def has_storage(self, address: int) -> bool:
+        pass
+
+    @abstractmethod
+    def get_storage(self, address: int) -> Optional[Storage]:
+        pass
+
+    # sam.moelius: The addition of the constraints parameter is future-proofing.  We might want
+    # to create a new Storage object in OverlayWorldState when self._storage.get(address) is None.
```

Comment:
In which case a Storage of a known address will be None?

Reply:
That sounds right.

Reply:
When the storage of an account that exists will be None? Is this for the special case of a remote world?

Reply:
Essentially, yes.